nemo/core/callbacks.py

```diff
@@ -35,6 +54,463 @@
+                every step_freq steps.
+                Defaults to ["loss"] which only prints the loss.
+        """
+        self.step_freq = step_freq
+        self.tensors_to_log = tensors_to_log
+
+    def on_step_end(self, state):
+        if state["step"] % self.step_freq == 0:
+            for tensor_key in self.tensors_to_log:
+                tensor = state["tensors"].get_tensor(tensor_key)
+                logging.info("%s: %s", tensor_key, tensor)
+
+
+class TensorboardLogger(NeMoCallback):
+    def __init__(
+        self,
+        tb_writer: 'torch.utils.tensorboard.SummaryWriter',
+        step_freq: int = 100,
+        tensors_to_log: List[Union[str, 'NmTensor']] = ["loss"],
+        custom_tb_log_func: Callable[[Union[str, 'NmTensor']], None] = None,
+        log_epoch: bool = True,
+    ):
+        """A tensorboard callback that logs tensors using a tensorboard writer object. It's default option is to log
+        the loss every 100 steps. Additional scalar tensors can be logged by adding them to the tensors_to_log
+        argument. In order to log complex tensorboard entities, the custom_tb_log_func must be passed it. By default,
+        it always logs the current epoch and the time taken per epoch.
+
+        args:
+            tb_writer (required): The tensorboard logger object.
+            step_freq (int): The frequency of tensorboard logging. Defaults to every 100 steps
+            tensors_to_log (List of str or NmTensor): A list of either tensor names or NmTensors which will be logged
+                every step_freq steps.
+                Defaults to ["loss"] which only prints the loss.
+            custom_tb_log_func (func): TensorboardLogger loops through tensors_to_log and passes these elements to
+                custom_tb_log_func. So a custom_tb_log_func will receive one argument on each call with the arugment
+                being an element from tensors_to_log.
+                Defaults to None which logs each tensors_to_log as a scalar.
+            log_epoch (bool): Whether to log epoch and epoch training time to tensorboard.
+                Defaults to True.
+        """
+        self.step_freq = step_freq
+        self.tensors_to_log = tensors_to_log
+        self.tb_writer = tb_writer
+        self.custom_tb_log_func = custom_tb_log_func
+        self._last_epoch_start = None
+        self._log_epoch = log_epoch
+
+    def on_epoch_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            self._last_epoch_start = time.time()
+
+    def on_epoch_end(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if self._log_epoch:
+                epoch_time = time.time() - self._last_epoch_start
+                self.tb_writer.add_scalar('misc/epoch', state["epoch"], state["step"])
+                self.tb_writer.add_scalar('misc/epoch_time', epoch_time, state["step"])
```

Comment:
Why do we have this inside of a TensorboardLogger?

This looks like a good, tiny TimeLogger callback to me... How about _decoupling_ those two functionalities?